### ISSUE TITLE
function pointer with pic

### ISSUE THREAD

**Issue by anjiahao1** [usvi/F070RB-BL-FW#3]
Hello,
I have a writing problem when using pic,
An error occurs when using a pointer to access a function in a global variable. The reason is that the function address in the global variable has been determined, but when using a pointer to access it, pic will not add the offset for me.

If you update the contents of .data during the system initialization phase, you cannot accurately distinguish which ones are absolute addresses and which ones are data.

example:
```
typedef void (*func)(void);

struct ops
{
  uint32_t a;
  func b;
};
void test(void)
{
  //do nothing
}

struct ops g_ops =
{
  12345,
  test
};

int main(void)
{
  struct ops *ops = &g_ops;

  ops->b();
  return 0;
}
```

do ops->b() will get a bad address.

**Comment by usvi** [comment]
Yes, notoriously function pointers are a problem at least with basic forms of PIC and should be avoided. Some globals are a problem also.

Dunno if you have read my writeup but here it is: https://techblog.paalijarvi.fi/2022/01/16/portable-position-independent-code-pic-bootloader-and-firmware-for-arm-cortex-m0-and-cortex-m4/

I don't know if these aspects can be solved in a proper way at all. One hack I have heard of was to put the function pointers to a struct and use somehow with that. But I don't remember/know the details.

I'm very sorry.

**Comment by anjiahao1** [comment]
Thank you for your reply. I found this code repository after reading your article. The article is very well written.
So for now, there is no way to access function pointers and certain global variables? Can such an operation be avoided only manually?

**Comment by anjiahao1** [comment]
i am do PIC on Nuttx rtos. just function pointer and some globals var is badaddress, other code is run well, i mark some function pointer，can run well.
However, mark all fuctions pointer is a verify lager work.
